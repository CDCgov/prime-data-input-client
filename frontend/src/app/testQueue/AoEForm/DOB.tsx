import Button from "../../commonComponents/Button";

const DOB = () => {
  const savePatientAnswers = () => {
    console.log("saved");
  };

  return (
    <>
      <main>
        <div className="grid-container maxw-tablet">
          <p className="margin-top-3">
            Enter your date of birth to access your COVID-19 Testing Portal.
          </p>
          <form className="usa-form">
            <label className="usa-label" htmlFor="bday" aria-describedby="bdayFormat">
              Date of birth
            </label>
            <span className="usa-hint text-ls-1">MM/DD/YYYY or MMDDYYYY</span>
            <input className="usa-input" id="bday" type="text" required aria-required="true" autoComplete="bday" size={8} pattern="[0-9]{8}" inputMode="numeric" />
            <Button label="Continue" onClick={savePatientAnswers} />
          </form>
        </div>
      </main>
    </>
  );
};

export default DOB;
