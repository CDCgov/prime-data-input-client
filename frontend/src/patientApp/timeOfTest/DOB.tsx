import React, { useEffect } from "react";
import { useState } from "react";
import moment from "moment";

import Button from "../../app/commonComponents/Button";
import TextInput from "../../app/commonComponents/TextInput";
import { gql, useLazyQuery } from "@apollo/client";
import { connect, useDispatch, useSelector } from "react-redux";
import { setPatient } from "../../app/store";
import { Redirect } from "react-router";

const PATIENT_LINK_VALIDATION_QUERY = gql`
  query PatientLinkVerify($plid: String!, $birthDate: LocalDate!) {
    patientLinkVerify(internalId: $plid, birthDate: $birthDate) {
      internalId
      firstName
      middleName
      lastName
      birthDate
      street
      streetTwo
      city
      state
      zipCode
      telephone
      role
      email
      county
      race
      ethnicity
      gender
      residentCongregateSetting
      employedInHealthcare
    }
  }
`;

const DOB = () => {
  const dispatch = useDispatch();

  const [birthDate, setBirthDate] = useState("");
  const [formattedBirthDate, setFormattedBirthDate] = useState("");
  const [birthDateError, setBirthDateError] = useState("");
  const [nextPage, setNextPage] = useState(false);

  const plid = useSelector((state) => (state as any).plid as String);

  const [validateBirthDate, { called, loading, data }] = useLazyQuery(
    PATIENT_LINK_VALIDATION_QUERY,
    {
      variables: { plid, birthDate: formattedBirthDate },
      fetchPolicy: "no-cache",
    }
  );

  useEffect(() => {
    if (!data) return;

    dispatch(setPatient(data.patientLinkVerify));

    setNextPage(true);
    // eslint-disable-next-line
  }, [data]);

  const isValidForm = () => {
    const validDate = moment(birthDate.replace("/", ""), "MMDDYYYY").isValid();

    if (validDate) {
      setBirthDateError("");
      return true;
    } else {
      setBirthDateError("Enter your date of birth");
      return false;
    }
  };

  const confirmBirthDate = () => {
    if (isValidForm()) {
      setFormattedBirthDate(
        moment(birthDate.replace("/", ""), "MMDDYYYY").format("YYYY-MM-DD")
      );
      validateBirthDate();
    }
  };

  if (called && loading) {
    return <p>Validating birth date...</p>;
  }

  if (nextPage) {
    return (
      <Redirect
        push
        to={{
          pathname: "/patient-info-confirmation",
          state: { page: "profile" },
        }}
      />
    );
  }

  return (
    <>
      <main>
        <div className="grid-container maxw-tablet">
          <p className="margin-top-3">
            Enter your date of birth to access your COVID-19 Testing Portal.
          </p>
          <form
            className="usa-form"
            onSubmit={(e) => {
              e.preventDefault();
              confirmBirthDate();
            }}
          >
            <TextInput
              label={"Date of birth"}
              name={"birthDate"}
              type={"bday"}
              required={true}
              autoComplete={"bday"}
              value={birthDate}
              size={8}
              pattern={"([0-9]{1,2}/[0-9]{1,2}/[0-9]{4})|([0-9]{8})"}
              inputMode={"numeric"}
              ariaDescribedBy={"bdayFormat"}
              hintText={"MM/DD/YYYY or MMDDYYYY"}
              errorMessage={birthDateError}
              validationStatus={birthDateError ? "error" : undefined}
              onChange={(evt) => setBirthDate(evt.currentTarget.value)}
            />
            <Button label={"Continue"} type={"submit"} />
          </form>
        </div>
      </main>
    </>
  );
};

export default connect()(DOB);
