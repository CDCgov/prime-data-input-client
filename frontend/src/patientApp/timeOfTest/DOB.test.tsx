import renderer from "react-test-renderer";
import { render, fireEvent, RenderResult } from "@testing-library/react";

import DOB from "./DOB";

const mockContainer = () => <DOB />;

jest.mock("../PxpApiService", () => {
  jest.fn();
});

describe("DOB", () => {
  describe("snapshot", () => {
    let component: renderer.ReactTestRenderer;
    beforeEach(() => {
      component = renderer.create(mockContainer());
    });
    it("matches", () => {
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe("behavior", () => {
    let utils: RenderResult;

    async function setup(utils: RenderResult) {
      const input = await utils.findByLabelText("Date of birth");
      const button = await utils.findByRole("button");
      return { input, button };
    }

    beforeEach(() => {
      utils = render(mockContainer());
    });

    it("validates birthdays", async () => {
      // GIVEN
      const { input, button } = await setup(utils);

      // WHEN
      fireEvent.change(input, { target: { value: "99-99-9999" } });
      fireEvent.submit(button);
      const error = await utils.findByRole("alert");

      // THEN
      expect(error.textContent).toEqual("Error: Enter your date of birth");
    });
  });
});
