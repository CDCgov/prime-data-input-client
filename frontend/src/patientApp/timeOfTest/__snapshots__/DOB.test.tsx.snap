// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOB snapshot matches 1`] = `
<main>
  <div
    className="grid-container maxw-tablet"
  >
    <p
      className="margin-top-3"
    >
      Enter your date of birth to access your COVID-19 Testing Portal.
    </p>
    <form
      className="usa-form"
      onSubmit={[Function]}
    >
      <div
        className="usa-form-group"
      >
        <label
          aria-describedby="bdayFormat"
          className="usa-label"
          htmlFor="1"
        >
          Date of birth
        </label>
        <span
          className="usa-hint text-ls-1"
        >
          MM/DD/YYYY or MMDDYYYY
        </span>
        <input
          aria-required="false"
          autoComplete="bday"
          className="usa-input"
          id="1"
          inputMode="numeric"
          name="birthDate"
          onChange={[Function]}
          pattern="([0-9]{1,2}/[0-9]{1,2}/[0-9]{4})|([0-9]{8})"
          size={8}
          type="password"
          value=""
        />
      </div>
      <button
        className="usa-button"
        id="dob-submit-button"
        type="submit"
      >
        Continue
      </button>
    </form>
  </div>
</main>
`;
